### PR TITLE
Feature/dsym

### PR DESCRIPTION
**Pull request by mitsuhiko**
First pass on dsym support.

**Do not merge yet**

### CODE REVIEW

File: src/sentry/utils/zip.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    rv = os.path.commonprefix(qualifying_members)
+    if rv:
+        return rv[0] + '/'
+    return ''
+
+
+def safe_extract_zip(f, path, strip_toplevel=True):
+    """Safely extract a given zip file to a path.  The zipfile can either
+    be an open file or a filename.  If the zip is unsafe an exception is
+    raised.  Optionally the toplevel fodler is stripped off.  If there are
```
Comment (dcramer) [inline src/sentry/utils/zip.py]: fodler

File: src/sentry/templates/sentry/projects/manage.html [hunk 2]
```diff
@@ -179,6 +179,9 @@ <h6 class="nav-header">{% trans "Configuration" %}</h6>
     <li>
         <a href="{% absolute_uri '/{}/{}/settings/saved-searches/' project.organization.slug project.slug %}">{% trans "Saved Searches" %}</a>
     </li>
+    <li>
```
Comment (dcramer) [inline src/sentry/templates/sentry/projects/manage.html]: might be worthwhile to put behind a feature switch

File: src/sentry/utils/zip.py (hunk 1 above)
Comment (mitsuhiko) [inline src/sentry/utils/zip.py]: Fixed.

File: src/sentry/interfaces/applecrash.py [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    >>> {
+    >>>     "crash": {...}
+    >>> }
+    """
+
+    ephemeral = True
+
+    @classmethod
+    def to_python(cls, data):
+        if not data.get('message'):
```
Comment (dcramer) [inline src/sentry/interfaces/applecrash.py]: this seems like a slip

File: src/sentry/interfaces/base.py [hunk 4]
```diff
@@ -7,6 +7,16 @@
 from sentry.utils.imports import import_string
 
 
+def iter_interfaces():
+    rv = {}
```
Comment (dcramer) [inline src/sentry/interfaces/base.py]: unused?

File: src/sentry/api/authentication.py [hunk 5]
```diff
@@ -16,6 +18,14 @@ def authenticate_header(self, request):
 
 class ApiKeyAuthentication(QuietBasicAuthentication):
     def authenticate_credentials(self, userid, password):
+        if userid == 'admin':
+            ref_pw = options.get('system.admin-auth-password')
```
Comment (dcramer) [inline src/sentry/api/authentication.py]: we could just do this as a username which is a "system API key"

File: src/sentry/lang/native/dsymcache.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+import shutil
+
+from django.conf import settings
+from django.db.models import Q
+from sentry.models import DSymFile
+
+
+class DSymCache(object):
+
+    def __init__(self):
```
Comment (dcramer) [inline src/sentry/lang/native/dsymcache.py]: this

File: src/sentry/lang/native/dsymcache.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
+import os
+import shutil
+
+from django.conf import settings
+from django.db.models import Q
+from sentry.models import DSymFile
```
Comment (dcramer) [inline src/sentry/lang/native/dsymcache.py]: linters/etc will yell about missing required import statement

File: src/sentry/lang/native/plugin.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        return
+
+    project = Project.objects.get_from_cache(
+        id=data['project'],
+    )
+
+    sym = make_symbolizer(project, crash_report['binary_images'])
+    crash = crash_report['crash']
+
+    bt = None
```
Comment (dcramer) [inline src/sentry/lang/native/plugin.py]: where are we storing threads, or are we ignoring those atm?

**Comment by dcramer** [comment]
So mostly seems sane, would like to see before merge:

- basic tests for the API endpoints (can probably just copy paste the release ones)
- fix various lint/imports
- put crash reports UI behind feature switch
- ensure that symsynd isnt required to run Sentry (maybe fail gracefully on loading sentry.lang.native, or conditionally install it)

File: src/sentry/lang/native/plugin.py (hunk 8 above)
Comment (mitsuhiko) [inline src/sentry/lang/native/plugin.py]: Currently threads are ignored because we have no interface for it yet.

**Pull request merged**